usr/src/boot/common/module.c

```diff
@@ -753,6 +773,15 @@ mod_load(char *modname, struct mod_depend *verinfo, int argc, char *argv[])
 	int err;
 	char *filename;
 
+	/*
+	 * On aarch64 systems the first loaded module must be the kernel.
```

Comment:
this is true on all cases actually. on x86, the check happens in file_loadraw() because in current implementation we do not load ELF, but raw file as is and dboot will copy text, data and bss in place. I have WIP to implement ELF loading instead, to be able to skip dboot at all (so we would not need to do 64-bit -> 32-bit -> 64-bit dance), and dboot would be there for some time to support other boot loaders.

Reply:
mm, ok, perhaps this is all right. I can not say now, I really haven't read all those its, just having experience from current x86 solution, where we have essentially 4 kernels - boot loader, then dboot, then locore (early kernel) and then final kernel. All with their implementations for communications etc. But as I wrote, if this approach does enable the things to work, it is good and can be improved later (if it does need it).